### DISCUSSION TITLE
Using batch.from_data_list with custom MessagePassing

### DISCUSSION THREAD

**Discussion by SmokeShine** [pyg-team/pytorch_geometric#9190] [Q&A] [answered]
Hi,

I am trying to understand how to use the batching process in PyG. I am able to create a batch of 10 different graphs using batch.from_data_list. However, they are stored as one big disconnected graph. 

I have created a custom MessagePassing layer and modified its forward with self.propagate and changed self.message and self.update.  As per my understanding,  when data is passed to layer, it will call forward like it does for pytorch, which will then do the aggregation using self.update and self.message. 

Now, I am not able to figure out how to use the batch of data that is coming. It is one big disconnected graph and not as a tensor of 10 graphs like it comes with pytorch.  I have the same index for nodes in these 10 graphs. Only the edge values changes, but the nodes are same.

So, I need help for two situation:
1. How to use the batch correctly with the messagepassing layer?
2. what to check for silently failing issues? what kind of unit test case that I can create?

I have searched the issue list on the github page, but I couldn't find a similar issue.

**Comment by rusty1s** [comment, marked as answer]
Message passing works by sending messages along the edges to update node representations. Since the graph is disconnected, there will be no message exchange between two disconnected graphs. As such, you don't need to worry about batching inside `MessagePassing` layer. This is described as best as I could [here](https://pytorch-geometric.readthedocs.io/en/latest/advanced/batching.html).

**Reply by SmokeShine** [reply]
Thanks @rusty1s  for the reply. 
Please let me know if my understanding is correct. I have a graph consisting of 5 customers and I create a batch of 10. So, internally, it will becomes a big disconnected graph of 50 customers. As the graphs are disconnected, the message passing from one graph will not influence another graph. So, in that case, the output will be 50 embeddings (one for each customer) and if I want to put a dense layer, I can use view function to get the batch as the first dimension and then call the usual pytorch layers.

**Reply by rusty1s** [reply]
That's correct.
